### PR TITLE
fix: resolve timezone bug in financial year date handling (SW-1109)

### PR DESCRIPTION
**Pull request by wheelsandcogs**
## Summary

Fixes SW-1109: financial year time period dates were displaying incorrectly due to timezone pollution.

**Root cause:** `date-fns/parse` created dates in the server's local timezone. When stored in `timestamp without time zone` columns, the UTC representation was timezone-shifted. The `pg` driver then reinterpreted values using local time on read, compounding the problem.

**Fix:**
- All date creation in `date-matching.ts` now uses UTC-explicit parsing (`parseISO` with `Z` suffix, `parseAsUTC` helper)
- All `format()` calls replaced with `formatUTC()` using `TZDate` from `@date-fns/tz` to ensure timezone-safe formatting regardless of server timezone
- Revision `start_date`/`end_date` columns migrated from `timestamp without time zone` to `date` — these are calendar dates, not moments in time
- Lookup table schema also changed to `DATE` columns
- Removed the `seconds: -1` hack for end dates — end dates are now clean calendar dates (e.g. `2024-03-31` instead of `2024-03-31 23:59:59`)
- Added TypeORM column transformers on `revision` and `dataset` entities so `date` columns are read back as proper Date objects (pg driver returns them as strings)
- Cleaned up the `typeof` guard in `consumer-dataset-dto.ts` that was working around this

## Test plan

- [ ] All 922 backend unit tests pass
- [ ] Rebuild a financial year dataset and verify `year_lookup` table has correct dates (e.g. `2001-04-01` not `2001-03-31 23:00:00`)
- [ ] Verify `revision.start_date` and `end_date` are correct after migration
- [ ] Check publisher dimension preview shows "1st April" for financial year start dates
- [ ] Check consumer About tab shows correct coverage period (e.g. "April 2013 to March 2024")

### CODE REVIEW

File: src/entities/dataset/dataset.ts [hunk 1]
```diff
@@ -52,10 +52,20 @@ export class Dataset extends BaseEntity {
   @Column({ name: 'replacement_auto_redirect', type: 'boolean', nullable: false, default: false })
   replacementAutoRedirect: boolean;
 
-  @Column({ name: 'start_date', type: 'date', nullable: true })
+  @Column({
+    name: 'start_date',
+    type: 'date',
+    nullable: true,
+    transformer: { to: (value: Date | null) => value, from: (value: string | null) => (value ? new Date(value) : null) }
+  })
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/entities/dataset/dataset.ts:66]: Fixed — the transformer `to` side now writes a `YYYY-MM-DD` string via `value.toISOString().split('T')[0]` instead of passing the Date object through.

File: src/entities/dataset/revision.ts [hunk 2]
```diff
@@ -122,9 +122,19 @@ export class Revision extends BaseEntity {
...
+  })
   startDate: Date | null;
 
-  @Column({ name: 'end_date', type: 'timestamp without time zone', nullable: true })
+  @Column({
+    name: 'end_date',
+    type: 'date',
+    nullable: true,
+    transformer: { to: (value: Date | null) => value, from: (value: string | null) => (value ? new Date(value) : null) }
+  })
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/entities/dataset/revision.ts:150]: Fixed — same change applied to revision.ts. Both entities now write `YYYY-MM-DD` strings on the `to` side.

File: src/migrations/1776249719058-revision-dates-to-date-type.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class RevisionDatesToDateType1776249719058 implements MigrationInterface {
+  name = 'RevisionDatesToDateType1776249719058';
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`ALTER TABLE "revision" ALTER COLUMN "start_date" TYPE date USING start_date::date`);
+    await queryRunner.query(`ALTER TABLE "revision" ALTER COLUMN "end_date" TYPE date USING end_date::date`);
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/migrations/1776249719058-revision-dates-to-date-type.ts:8]: Prod runs UTC so the stored values are clean midnight timestamps — `::date` truncation produces the correct calendar date. Local dev data created under BST would be wrong, but that's throwaway. Not adding the `+ INTERVAL '12 hours'` trick as it could overcorrect edge cases on UTC data.

File: test/services/date-matching.test.ts [hunk 4]
```diff
@@ -102,53 +102,53 @@ describe('date-matching', () => {
     test('Calendar year starts Jan 1 and ends Dec 31', () => {
       const extractor: DateExtractor = { type: YearType.Calendar, yearFormat: 'YYYY' };
       const result = dateDimensionReferenceTableCreator(extractor, [{ dateCode: '2023' }]);
-      const expectedStart = parse('2023-01-01', 'yyyy-MM-dd', new Date());
-      const expectedEnd = add(expectedStart, { months: 12, seconds: -1 });
+      const expectedStart = parseISO('2023-01-01T00:00:00Z');
+      const expectedEnd = sub(add(expectedStart, { months: 12 }), { days: 1 });
       expect(result[0].start).toEqual(expectedStart);
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply test/services/date-matching.test.ts]: Added 9 timezone-independence tests that run under `America/New_York`, `Asia/Kolkata`, and `Pacific/Auckland`, covering financial year, point-in-time, and rolling period dates. Assertions use `Date.UTC()` for explicit UTC epoch comparison.

[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply test/services/date-matching.test.ts]: Fixed — test expected values now use explicit UTC dates (`utc()` helper wrapping `TZDate`) or `Date.UTC()` rather than local-time `add()`/`sub()` arithmetic.

File: src/services/date-matching.ts [hunk 5]
```diff
@@ -41,8 +42,8 @@ export const createDatePeriodTableQuery = (
         %I %s,
         language VARCHAR(5),
         description VARCHAR,
-        start_date TIMESTAMP WITHOUT TIME ZONE,
-        end_date TIMESTAMP WITHOUT TIME ZONE,
+        start_date DATE,
+        end_date DATE,
         date_type VARCHAR,
         sort_order BIGINT,
         hierarchy %s
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/services/date-matching.ts:49]: Fixed — dates are now formatted as `YYYY-MM-DD` strings (via `toISOString().split('T')[0]`) before being passed to the INSERT statements and CSV serialization, so pg receives plain date literals rather than timestamps.

File: src/services/date-matching.ts [hunk 6]
```diff
@@ -55,6 +56,18 @@ export const createDatePeriodTableQuery = (
   );
 };
 
+// Parse a date string into a UTC midnight Date, avoiding local-timezone pollution.
+// Uses date-fns parse to understand the format, then reconstructs as UTC.
+function parseAsUTC(value: string, formatStr: string): Date {
+  const parsed = parse(value, formatStr, new Date());
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/services/date-matching.ts:69]: Fixed — `parseAsUTC` now checks `isValid(parsed)` and returns the plain invalid Date unchanged, so downstream `isValid()` checks still work cleanly.

File: src/services/date-matching.ts [hunk 7]
```diff
@@ -277,8 +290,8 @@ function periodTableCreator(
   const endYear = Math.max(...dataYears);
   const type = yearType(dateFormat.type, dateFormat.startDay, dateFormat.startMonth);
 
-  let year = parse(`${startYear}-${type.start}`, 'yyyy-MM-dd', new Date());
-  const end = add(parse(`${endYear}-${type.start}`, 'yyyy-MM-dd', new Date()), { years: 1, seconds: -1 });
+  let year = parseISO(`${startYear}-${type.start}T00:00:00Z`);
+  const end = add(parseISO(`${endYear}-${type.start}T00:00:00Z`), { years: 1 });
   // Quarters and month numbers are different depending on the type of year
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/services/date-matching.ts]: Fixed — added `toUTC()` and `parseISOasUTC()` helpers that wrap dates as `TZDate('UTC')` at parse time. All downstream `add()`/`sub()`/`isBefore()` calls now operate on `TZDate` objects, so arithmetic is performed in UTC regardless of server timezone.

File: src/entities/dataset/dataset.ts [hunk 8]
```diff
@@ -52,10 +52,20 @@ export class Dataset extends BaseEntity {
...
   startDate: Date | null;
 
-  @Column({ name: 'end_date', type: 'date', nullable: true })
+  @Column({
+    name: 'end_date',
+    type: 'date',
+    nullable: true,
+    transformer: { to: (value: Date | null) => value, from: (value: string | null) => (value ? new Date(value) : null) }
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/entities/dataset/dataset.ts]: Already fixed in the latest push — both `to` sides write `YYYY-MM-DD` strings.

File: src/entities/dataset/revision.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/entities/dataset/revision.ts:150]: Already fixed in the latest push.

File: src/services/dimension-processor.ts [hunk 9]
```diff
@@ -754,8 +754,8 @@ async function getDatePreviewWithExtractor(
       SELECT
           %I.%I as %I,
           %I.description,
-          to_char(%I.start_date, 'YYYY-MM-DD"T"HH24:MI:SS"Z"') as start_date,
-          to_char(%I.end_date, 'YYYY-MM-DD"T"HH24:MI:SS"Z"') as end_date,
+          to_char(%I.start_date, 'YYYY-MM-DD') as start_date,
+          to_char(%I.end_date, 'YYYY-MM-DD') as end_date,
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/services/dimension-processor.ts:758]: The inserts now use `toISOString().split('T')[0]` to write `YYYY-MM-DD` strings. The `to_char` here is on the read/SELECT path — the column is already `DATE`, so the `::date` cast is redundant.

File: src/dtos/consumer-dataset-dto.ts [hunk 10]
```diff
@@ -46,8 +46,8 @@ export class ConsumerDatasetDTO {
       dto.published_revision = ConsumerRevisionDTO.fromRevision(dataset.publishedRevision);
     }
 
-    dto.start_date = typeof dataset.startDate === 'string' ? dataset.startDate : dataset.startDate?.toISOString();
-    dto.end_date = typeof dataset.endDate === 'string' ? dataset.endDate : dataset.endDate?.toISOString();
+    dto.start_date = dataset.startDate?.toISOString();
+    dto.end_date = dataset.endDate?.toISOString();
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/dtos/consumer-dataset-dto.ts]: Good point — updated all four consumer DTOs to return `toISOString().split('T')[0]` (date-only `YYYY-MM-DD`) for start/end dates.

File: src/entities/dataset/revision.ts [hunk 11]
```diff
@@ -122,9 +122,19 @@ export class Revision extends BaseEntity {
   })
   cubeType: CubeType | null;
 
-  @Column({ name: 'start_date', type: 'timestamp without time zone', nullable: true })
+  @Column({
+    name: 'start_date',
+    type: 'date',
+    nullable: true,
+    transformer: { to: (value: Date | null) => value, from: (value: string | null) => (value ? new Date(value) : null) }
+  })
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/entities/dataset/revision.ts:136]: Already fixed in the latest push.

File: src/services/date-matching.ts [hunk 12]
```diff
@@ -498,9 +511,9 @@ function periodDateTableCreator(dateFormat: DateExtractor, dataColumn: string[])
     }
 
     for (const locale of SUPPORTED_LOCALES) {
-      const fullDate = `${parsedDate.getDate()} ${t(`months.${parsedDate.getMonth() + 1}`, { lng: locale })} ${parsedDate.getFullYear()}`;
+      const fullDate = `${parsedDate.getUTCDate()} ${t(`months.${parsedDate.getUTCMonth() + 1}`, { lng: locale })} ${parsedDate.getUTCFullYear()}`;
       const startDate = sub(add(parsedDate, { days: 1 }), type.increment);
-      const endDate = add(parsedDate, { days: 1, seconds: -1 });
+      const endDate = parsedDate;
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply src/services/date-matching.ts:531]: Already fixed — `parseAsUTC` and `parseISOasUTC` return `TZDate('UTC')` objects, so all downstream `add`/`sub` calls operate in UTC.

File: test/services/date-matching.test.ts [hunk 13]
```diff
@@ -102,53 +102,53 @@ describe('date-matching', () => {
...
     });
 
     test('Financial year starts Apr 1 and ends Mar 31', () => {
       const extractor: DateExtractor = { type: YearType.Financial, yearFormat: 'YYYYYY' };
       const result = dateDimensionReferenceTableCreator(extractor, [{ dateCode: '202324' }]);
-      const expectedStart = parse('2023-04-01', 'yyyy-MM-dd', new Date());
-      const expectedEnd = add(expectedStart, { months: 12, seconds: -1 });
+      const expectedStart = parseISO('2023-04-01T00:00:00Z');
+      const expectedEnd = sub(add(expectedStart, { months: 12 }), { days: 1 });
       expect(result[0].start).toEqual(expectedStart);
```
[the remark this answers is not in the document]
Comment (wheelsandcogs) [reply test/services/date-matching.test.ts]: Already fixed — 9 timezone-independence tests added across 3 non-UTC timezones, and test expectations use explicit UTC values.

**Review by j-maynard — APPROVED** [review]
LGTM 🤘🏻
